### DISCUSSION TITLE
Drums on channel different than 10 for GS soundfont

### DISCUSSION THREAD

**Discussion by lorenzosu** [FluidSynth/fluidsynth#1326] [Q&A]
Hi all,
For GS type soundfonts synths there were/are ways to force a channel different than 10 to do drumkits either via sets of controllers or sysex with the effect of being able to use multiple drumkits at the same time. Is this possible emulate this behaviour with FluidSynth?
The scenario would, for example, 'mixing' drumkits present in the same soundfont (inculding for example the Fluid one).
Thanks

**Comment by derselbst** [comment]
Yes, since fluidsynth 2.2.0 we support GS DT1 SysEx messages for this purpose, see #708.

Implementation goes here:

https://github.com/FluidSynth/fluidsynth/blob/f9bde410216766c1117cd07b349740e9804773ca/src/synth/fluid_synth.c#L2443-L2467

**Comment by lorenzosu** [comment]
Hi thanks for the explanation. Is there a working example of how one would set this up to work with a soundfont such as FluidR3_GM_GS.sf2?

I tried the midi file cited #708  but it didn't seem to work as expected even with setting the device id to 16 (i.e. piano playing on channel 10 and drums on channel 2)

I also tried using the example in the cited SC-88Pro/8850 i.e. sending `41 10 42 12 40 1A 15 02 0F F7` with the `synth.device-id` set to 16 but then FluidSynth reports:
`fluidsynth: SysEx DT1: dropping message on addr 0x401a15 due to incorrect checksum 0xfffffff7. Correct checksum: 0x80`

I tried calculating the checksum as described in the Roland manual and that should be the correct value.

I'm wondering if a simpler way to implement this with a (fluidsynth 'native') message which in could be something like 'set X channel to like a drum one, exactly as for channel 10' in more human / musical terms.

**Reply by derselbst** [reply]
I currently have no time to provide you with  a working SysEx example, sry. You can try pinging the author of that ticket.

Alternatively, you can use [fluid_synth_set_channel_type()](https://www.fluidsynth.org/api/group__channel__setup.html#gae6de53f3095bb7a68c53dbd5c914a9fb).

**Comment by lorenzosu** [comment]
OK thanks. can  fluid_synth_set_channel_type() be somehow set interactively via the fluidsynth (commandline) runtime?

**Reply by derselbst** [reply]
No, we don't offer a shell command for this currently, sry.

**Reply by lorenzosu** [reply]
Hi I actually tried to add this in `fluid_synth.c` by creating a new setting "synth.midi-drum-channels" which would take a comma-separated list of channels (this is mimicking the setting "synth.overflow.important-channels").

Idea is that via changing this setting the [fluid_synth_set_channel_type()](https://www.fluidsynth.org/api/group__channel__setup.html#gae6de53f3095bb7a68c53dbd5c914a9fb) function is called on the specified channels setting them to `CHANNEL_TYPE_DRUM` effectively allowing to play drums on channels different than 10

Specifying a setting of (say) -1 would 'reset' the setting and go back to channel 10 only.

I'm not sure if a setting like this would be acceptable (in my case I'd find it quite useful) and also if this implementation would.

Modified file is here: [fluid_synth.c ](https://gist.github.com/lorenzosu/90ac2784991bd96bf86c5d77c1802bdb#file-fluid_synth-c)
Diff is here: https://gist.github.com/lorenzosu/d189edc30e0c40587c4c8195b5c18f1b

Here is also a quick demo video: 

https://github.com/FluidSynth/fluidsynth/assets/463937/acc0c883-66fe-49a3-bc4f-52ce9d07e326

**Reply by derselbst** [reply]
Sorry for the late response. That looks quite good, thanks! You're welcome to file a PR for that.

One thought: Does it make sense for the setting to be changed in realtime? Because if you do a `set`, it does apply the changes to the synth. However drum and melodic channels and also be changed by MIDI SysEX events or through the API. But those changes will not be reflected in the `synth.midi-drum-channels` settings string. Leading to inconsistencies if you `set synth.midi-drum-channels` at a later point.

**Reply by notaf0e** [reply]
Is there a way to expose the change channel type function to a python script? I attempted to use `ctypes` with `pyfluidsynth` but it didn't work. The expected behavior is that when using the `select` command to change banks and program change on a drum channel that it should only allow bank=128 but my attempt didn't produce this sadly
